### DISCUSSION TITLE
Handling Concurrency while sharing the session

### DISCUSSION THREAD

**Discussion by ccrvlh** [sqlalchemy/sqlalchemy#8554] [Usage Questions]
As we increase our async usage throughout the application and stop handling session lifecycle from within methods, we started having a few issues with concurrent operations (using `asyncio.gather(*tasks)` where tasks share the same session) - We have been getting a lot of `InterfaceErrors: cannot perform operation: another operation is in progress` I think I understand the reasons (same session, concurrent), but I couldn't find a clean way to handle this. What I thought so far:

- Go back to handling sessions within methods: each method would open/close their own session. Could work, I would lose the unit of work, but would always make sure the session is unique within a method.
- Start using the connection more often: we already use it `conn = await session.connection()` and then `results = await conn.execute(query)`, but we normally use it when we don't want the Models to be returned, but prefer the namedtuples/Rows (this is the case when we will be loading the result to a Dataframe for example). As I understand the connection always grabs/releases a new connection from the Pool, so In a way, the lifecycle is also being handled within the methods and wouldn't present an issue for concurrent logic (I think?)
- Is there anyway to "refresh" (get a new) session from within an existing session? I mean if I instantiate a class that received AsyncSession in the constructor, is there a way to get a new session using this object?
- Instead of passing the `AsyncSession` around, start passing the `sessionmaker`, I imagine this would a similar effect as (1), but not sure.

Am I approaching this the right way? Does the reasoning makes sense?

I saw a discussion for this topic where zzzeek talked about concurrency, but I just can't find it, I'll update this if I do.
I sent this message in the Gitter community, as @CaselIT asked I'm opening up the discussion, could probably make it easier for others to participate and contribute on this.

Thanks!

**Comment by zzzeek** [comment]
overall you would need the creation of the Session to be at the top of where a new task is created.  so if you have `asyncio.gather(*tasks)`, each of those tasks would have their own Session.  if you are trying to do something where the state of one lead Session is updated from the result of that gather operation, we'd need to look into what exactly you're trying to do transaction-wise.    If you are trying to work just in one transaction, that transaction does not accommodate concurrent IO and it doesnt make any sense that this would even be possible since transactions are a serial sequence of steps.  you would not be using `asyncio.gather(*tasks)` local to one transaction.

**Comment by CaselIT** [comment]
Thanks for posting here.
As mentioned in gitter, you should decide what tasks need to be in the same transaction and what do not.
For the one that require to be in the same transaction you would need to serialized them, since a single session cannot be used concurrently. For the other task you would instead use a separate session.

Personally I think you should manage the session outside of the task, and maybe use a helper function to create a new session and pass it to the task. something like:
```py
async def run_with_session(self, coroutine, *args, **kw):
  async with session_maker() as session:
    return await coroutine(*args, **kw, session=session)
```

> Start using the connection more often: we already use it conn = await session.connection() and then results = await conn.execute(query), but we normally use it when we don't want the Models to be returned, but prefer the namedtuples/Rows (this is the case when we will be loading the result to a Dataframe for example). As I understand the connection always grabs/releases a new connection from the Pool, so In a way, the lifecycle is also being handled within the methods and wouldn't present an issue for concurrent logic (I think?)

No, once a session has been associated with a given connection always the same is returned. You have to close a session for it to grab another connection.

**Comment by ccrvlh** [comment]
Thanks @CaselIT  and @zzzeek .

The intent was not really to have a single transaction, this ended up happening more as a side effect of having the session being injected in a services (`class AbcServices`), and having methods that could run concurrently (from a business logic perspective). The reason they end up with the same session is purely because they happen in the same web request (we've been trying to implement the best practices recommended by the docs). We hadn't encountered this challenge before since we were opening/closing session within methods, but for a lot of reasons we are much happier having the session being injected.

If the connection method returns the same connection (tied to the session it's being called from), it does seem that having a `RunWithSession` wrapper is probably the way to go, we were passing the coroutine (from a class that had the session injected) directly as a task to `asyncio.gather`, but by wrapping them in a helper function as @CaselIT mentioned would allows us to pass different session for different methods and still get the concurrency benefits.

I wonder if would make sense to have a helper method in the session object that would act more like a factory, maybe a `refresh_session()` or something, but that's just something I though for this specific use case, can't really think in any other use case for that.

Just for the sake of clarity, we usually have something like this now:
```python
class ModuleService: # this is what we are calling "the service"
    def __init__(self, session: AsyncSession, ...): # this is what I mean by "having the session injected in the service"
        self.session = session

    # this is what I'm calling the coroutine or service's method
    async def module_method(self,...):
        query = ...
        results = await self.session.execute(query)
        ...
```

**Reply by CaselIT** [reply]
Understood. Yes that pattern is fine if the operation of ModuleService are serial (ie you are not using `gather`). Likewise the suggestion of using a single session for web-request is fine if you are not planning on doing concurrent db operations in a that single request.
